File: app/src/main/java/com/example/schedule/compose/repo/TempScheduleRepo.java
/*
package com.example.schedule.compose.repo;

import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.example.schedule.ScheduleDBHelper;
import com.example.schedule.entity.Flow;
import com.example.schedule.entity.Lesson;
import com.example.schedule.entity.TempSchedule;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class TempScheduleRepo {
    private final ScheduleDBHelper dbHelper;
    private final FlowRepo flowRepo;
    private final LessonRepo lessonRepo;

    public TempScheduleRepo(Context context) {
        this.dbHelper = new ScheduleDBHelper(context);
        this.flowRepo = new FlowRepo(context);
        this.lessonRepo = new LessonRepo(context);
    }

    public TempScheduleRepo(ScheduleDBHelper dbHelper, FlowRepo flowRepo, LessonRepo lessonRepo) {
        this.dbHelper = dbHelper;
        this.flowRepo = flowRepo;
        this.lessonRepo = lessonRepo;
    }

    public long add(
            long flow, long lesson, LocalDate lessonDate, int lessonNum, boolean willLessonBe
    ) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        ContentValues values = new ContentValues();
        values.put("flow", flow);
        values.put("lesson", lesson);
        values.put(
                "lesson_date",
                lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
        );
        values.put("lesson_num", lessonNum);
        values.put("will_lesson_be", willLessonBe ? "1" : "0");

        long id = db.insert("temp_schedule", null, values);
        db.close();
        return id;
    }

    public long add(
            int flowLvl, int course, int flow, int subgroup,
            String name, String teacher, String cabinet,
            LocalDate lessonDate, int lessonNum, boolean willLessonBe
    ) {
        long flowId = -1;
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow != null) flowId = foundFlow.getId();
        else flowId = flowRepo.add(flowLvl, course, flow, subgroup);

        long lessonId = -1;
        Lesson foundLesson = lessonRepo.findByNameAndTeacherAndCabinet(
                name, teacher, cabinet
        );
        if (foundLesson != null) lessonId = foundLesson.getId();
        else lessonId = lessonRepo.add(name, teacher, cabinet);

        return add(flowId, lessonId, lessonDate, lessonNum, willLessonBe);
    }

    public int update(
            long flow, long lesson, LocalDate lessonDate, int lessonNum, boolean willLessonBe
    ) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        ContentValues values = new ContentValues();
        values.put("lesson", lesson);
        values.put("will_lesson_be", willLessonBe ? "1" : "0");

        String whereClause = "flow=? AND lesson_date=? AND lesson_num=?";
        String[] whereArgs = new String[] {
                String.valueOf(flow), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")),
                String.valueOf(lessonNum)
        };

        int count = db.update("temp_schedule", values, whereClause, whereArgs);
        db.close();
        return count;
    }

    public int update(
            int flowLvl, int course, int flow, int subgroup,
            String name, String teacher, String cabinet,
            LocalDate lessonDate, int lessonNum, boolean willLessonBe
    ) {
        long flowId = -1;
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow != null) flowId = foundFlow.getId();
        else flowId = flowRepo.add(flowLvl, course, flow, subgroup);

        long lessonId = -1;
        Lesson foundLesson = lessonRepo.findByNameAndTeacherAndCabinet(
                name, teacher, cabinet
        );
        if (foundLesson != null) lessonId = foundLesson.getId();
        else lessonId = lessonRepo.add(name, teacher, cabinet);

        return update(flowId, lessonId, lessonDate, lessonNum, willLessonBe);
    }

    public void addOrUpdate(
            long flow, long lesson, LocalDate lessonDate, int lessonNum, boolean willLessonBe
    ) {
        TempSchedule foundSchedule = findByFlowAndLessonDateAndLessonNum(
                flow, lessonDate, lessonNum
        );
        if (foundSchedule != null) update(flow, lesson, lessonDate, lessonNum, willLessonBe);
        else add(flow, lesson, lessonDate, lessonNum, willLessonBe);
    }

    public void addOrUpdate(
            int flowLvl, int course, int flow, int subgroup,
            String name, String teacher, String cabinet,
            LocalDate lessonDate, int lessonNum, boolean willLessonBe
    ) {
        long flowId = -1;
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow != null) flowId = foundFlow.getId();
        else flowId = flowRepo.add(flowLvl, course, flow, subgroup);

        long lessonId = -1;
        Lesson foundLesson = lessonRepo.findByNameAndTeacherAndCabinet(
                name, teacher, cabinet
        );
        if (foundLesson != null) lessonId = foundLesson.getId();
        else lessonId = lessonRepo.add(name, teacher, cabinet);

        addOrUpdate(flowId, lessonId, lessonDate, lessonNum, willLessonBe);
    }

    @SuppressLint("Range")
    public TempSchedule findByFlowAndLessonDateAndLessonNum(
            long flow, LocalDate lessonDate, int lessonNum
    ) {
        TempSchedule result = null;
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        String sql = "SELECT * FROM temp_schedule " +
                "WHERE flow=? AND lesson_date=? AND lesson_num=?;";
        String[] selectionArgs = new String[] {
                String.valueOf(flow), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")),
                String.valueOf(lessonNum)
        };
        Cursor cursor = db.rawQuery(sql, selectionArgs);
        if (cursor.moveToFirst()) {
            long id = cursor.getLong(cursor.getColumnIndex("id"));
            long lesson = cursor.getLong(cursor.getColumnIndex("lesson"));
            boolean willLessonBe = cursor.getInt(
                    cursor.getColumnIndex("will_lesson_be")
            ) == 1;

            result = new TempSchedule();
            result.setId(id);
            result.setFlow(flow);
            result.setLesson(lesson);
            result.setLessonDate(lessonDate);
            result.setLessonNum(lessonNum);
            result.setWillLessonBe(willLessonBe);
        }
        cursor.close();
        db.close();
        return result;
    }

    @SuppressLint("Range")
    public TempSchedule findByFlowAndLessonDateAndLessonNum(
            int flowLvl, int course, int flow, int subgroup, LocalDate lessonDate, int lessonNum
    ) {
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow == null) return null;

        long flowId = foundFlow.getId();

        return findByFlowAndLessonDateAndLessonNum(flowId, lessonDate, lessonNum);
    }

    @SuppressLint("Range")
    public List<TempSchedule> findAllByFlow(long flow) {
        List<TempSchedule> result = new ArrayList<>();
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        String sql = "SELECT * FROM temp_schedule WHERE flow=?;";
        String[] selectionArgs = new String[] {String.valueOf(flow)};
        Cursor cursor = db.rawQuery(sql, selectionArgs);
        while (cursor.moveToNext()) {
            long id = cursor.getLong(cursor.getColumnIndex("id"));
            long lesson = cursor.getLong(cursor.getColumnIndex("lesson"));
            LocalDate lessonDate = LocalDate.parse(
                    cursor.getString(cursor.getColumnIndex("lesson_date")),
                    DateTimeFormatter.ofPattern("yyyy-MM-dd")
            );
            int lessonNum = cursor.getInt(cursor.getColumnIndex("lesson_num"));
            boolean willLessonBe = cursor.getInt(
                    cursor.getColumnIndex("will_lesson_be")
            ) == 1;

            TempSchedule tempSchedule = new TempSchedule();
            tempSchedule.setId(id);
            tempSchedule.setFlow(flow);
            tempSchedule.setLesson(lesson);
            tempSchedule.setLessonDate(lessonDate);
            tempSchedule.setLessonNum(lessonNum);
            tempSchedule.setWillLessonBe(willLessonBe);

            result.add(tempSchedule);
        }
        cursor.close();
        db.close();
        return result;
    }

    public List<TempSchedule> findAllByFlow(int flowLvl, int course, int flow, int subgroup) {
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow == null) return null;

        long flowId = foundFlow.getId();

        return findAllByFlow(flowId);
    }

    public int deleteByFlowAndLessonDateAndLessonNum(
            long flow, LocalDate lessonDate, int lessonNum
    ) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        String whereClause = "flow=? AND lesson_date=? AND lesson_num=?";
        String[] whereArgs = new String[] {
                String.valueOf(flow), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")),
                String.valueOf(lessonNum)
        };

        int count = db.delete("temp_schedule", whereClause, whereArgs);
        db.close();
        return count;
    }

    public int deleteByFlowAndLessonDateAndLessonNum(
            int flowLvl, int course, int flow, int subgroup, LocalDate lessonDate, int lessonNum
    ) {
        Flow foundFlow = flowRepo.findByFlowLvlAndCourseAndFlowAndSubgroup(
                flowLvl, course, flow, subgroup
        );
        if (foundFlow == null) return 0;

        long flowId = foundFlow.getId();

        return deleteByFlowAndLessonDateAndLessonNum(flowId, lessonDate, lessonNum);
    }

    public int deleteAllBeforeDate(LocalDate date) {
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        String whereClause = "lesson_date<?";
        String[] whereArgs = new String[] {
                date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
        };

        int count = db.delete("temp_schedule", whereClause, whereArgs);
        db.close();
        return count;
    }
}
*/

File: app/src/main/java/ru/khanin/dmitrii/schedule/compose/repo/TempScheduleRepo.kt
package ru.khanin.dmitrii.schedule.compose.repo

import android.annotation.SuppressLint
import android.content.ContentValues
import android.database.Cursor
import ru.khanin.dmitrii.schedule.compose.entity.Cabinet
import ru.khanin.dmitrii.schedule.compose.entity.Flow
import ru.khanin.dmitrii.schedule.compose.entity.Subject
import ru.khanin.dmitrii.schedule.compose.entity.Teacher
import ru.khanin.dmitrii.schedule.compose.entity.TempSchedule
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class TempScheduleRepo(
    private val dbHelper: ScheduleDBHelper,
    private val flowRepo: FlowRepo,
    private val subjectRepo: SubjectRepo,
    private val teacherRepo: TeacherRepo,
    private val cabinetRepo: CabinetRepo
) {
    fun add(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        lessonDate: LocalDate,
        lessonNum: Int,
        willLessonBe: Boolean,
        subject: String?,
        surname: String?,
        name: String?,
        patronymic: String?,
        cabinet: String?,
        building: String?
    ): Long {
        val foundFlow = flowRepo.findByEducationLevelAndCourseAndGroupAndSubgroup(educationLevel, course, group, subgroup)
        val flowId = foundFlow?.id ?: flowRepo.add(educationLevel, course, group, subgroup)

        var subjectId: Long? = null
        if (subject?.isNotBlank() == true) {
            val foundSubject = subjectRepo.findBySubject(subject)
            subjectId = foundSubject?.id ?: subjectRepo.add(subject)
        }

        var teacherId: Long? = null
        if (surname?.isNotBlank() == true) {
            val foundTeacher = teacherRepo.findBySurnameAndNameAndPatronymic(surname, name, patronymic)
            teacherId = foundTeacher?.id ?: teacherRepo.add(surname, name, patronymic)
        }

        var cabinetId: Long? = null
        if (cabinet?.isNotBlank() == true) {
            val foundCabinet = cabinetRepo.findByCabinetAndBuilding(cabinet, building)
            cabinetId = foundCabinet?.id ?: cabinetRepo.add(cabinet, building)
        }

        val values = ContentValues()
        values.put("flow", flowId)
        values.put("lesson_date", lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")))
        values.put("lesson_num", lessonNum)
        values.put("will_lesson_be", if (willLessonBe) "1" else "0")
        values.put("subject", subjectId)
        values.put("teacher", teacherId)
        values.put("cabinet", cabinetId)

        val db = dbHelper.writableDatabase
        val id = db.insert("temp_schedule", null, values)
        db.close()
        return id
    }

    fun update(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        lessonDate: LocalDate,
        lessonNum: Int,
        willLessonBe: Boolean,
        subject: String?,
        surname: String?,
        name: String?,
        patronymic: String?,
        cabinet: String?,
        building: String?
    ): Int {
        val foundFlow = flowRepo.findByEducationLevelAndCourseAndGroupAndSubgroup(educationLevel, course, group, subgroup)
        val flowId = foundFlow?.id ?: flowRepo.add(educationLevel, course, group, subgroup)

        var subjectId: Long? = null
        if (subject?.isNotBlank() == true) {
            val foundSubject = subjectRepo.findBySubject(subject)
            subjectId = foundSubject?.id ?: subjectRepo.add(subject)
        }

        var teacherId: Long? = null
        if (surname?.isNotBlank() == true) {
            val foundTeacher = teacherRepo.findBySurnameAndNameAndPatronymic(surname, name, patronymic)
            teacherId = foundTeacher?.id ?: teacherRepo.add(surname, name, patronymic)
        }

        var cabinetId: Long? = null
        if (cabinet?.isNotBlank() == true) {
            val foundCabinet = cabinetRepo.findByCabinetAndBuilding(cabinet, building)
            cabinetId = foundCabinet?.id ?: cabinetRepo.add(cabinet, building)
        }

        val values = ContentValues()
        values.put("will_lesson_be", if (willLessonBe) "1" else "0")
        values.put("subject", subjectId)
        values.put("teacher", teacherId)
        values.put("cabinet", cabinetId)

        val whereClause = "flow=? AND lesson_date=? AND lesson_num=?"
        val whereArgs = arrayOf(flowId.toString(), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")), lessonNum.toString())

        val db = dbHelper.writableDatabase
        val count = db.update("temp_schedule", values, whereClause, whereArgs)
        db.close()
        return count
    }

    fun addOrUpdate(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        lessonDate: LocalDate,
        lessonNum: Int,
        willLessonBe: Boolean,
        subject: String?,
        surname: String?,
        name: String?,
        patronymic: String?,
        cabinet: String?,
        building: String?
    ) {
        val foundTempSchedule = findByFlowAndLessonDateAndLessonNum(educationLevel, course, group, subgroup, lessonDate, lessonNum)
        foundTempSchedule?.let {
            update(educationLevel, course, group, subgroup, lessonDate, lessonNum, willLessonBe, subject, surname, name, patronymic, cabinet, building)
        } ?: run {
            add(educationLevel, course, group, subgroup, lessonDate, lessonNum, willLessonBe, subject, surname, name, patronymic, cabinet, building)
        }
    }

    @SuppressLint("Range")
    fun findByFlowAndLessonDateAndLessonNum(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        lessonDate: LocalDate,
        lessonNum: Int
    ): TempSchedule? {
        var result: TempSchedule? = null

        val sql = """
            SELECT ts.id AS temp_id, ts.flow AS flow_id, ts.subject AS subject_id, ts.teacher AS teacher_id, ts.cabinet AS cabinet_id,
                ts.will_lesson_be, s.subject, t.surname, t.name, t.patronymic, c.cabinet, c.building, c.address
            FROM temp_schedule ts
            JOIN flow f ON ts.flow = f.id
            LEFT JOIN subject s ON ts.subject = s.id
            LEFT JOIN teacher t ON ts.teacher = t.id
            LEFT JOIN cabinet c ON ts.cabinet = c.id
            WHERE education_level=? AND course=? AND _group=? AND subgroup=? AND lesson_date=? AND lesson_num=?
        """.trimIndent()
        val selectionArgs = arrayOf(educationLevel.toString(), course.toString(), group.toString(), subgroup.toString(), lessonDate.toString(), lessonNum.toString())
        val db = dbHelper.readableDatabase
        val cursor: Cursor = db.rawQuery(sql, selectionArgs)
        if (cursor.moveToFirst()) {
            val flowId = cursor.getLong(cursor.getColumnIndex("flow_id"))

            val subjectId = cursor.getLong(cursor.getColumnIndex("subject_id"))
            val subject = cursor.getString(cursor.getColumnIndex("subject"))

            val teacherId = cursor.getLong(cursor.getColumnIndex("teacher_id"))
            val surname = cursor.getString(cursor.getColumnIndex("surname"))
            val name = cursor.getString(cursor.getColumnIndex("name"))
            val patronymic = cursor.getString(cursor.getColumnIndex("patronymic"))

            val cabinetId = cursor.getLong(cursor.getColumnIndex("cabinet_id"))
            val cabinet = cursor.getString(cursor.getColumnIndex("cabinet"))
            val building = cursor.getString(cursor.getColumnIndex("building"))
            val address = cursor.getString(cursor.getColumnIndex("address"))

            val id = cursor.getLong(cursor.getColumnIndex("temp_id"))
            val willLessonBe = cursor.getInt(cursor.getColumnIndex("will_lesson_be")) == 1

            result = TempSchedule(
                id,
                Flow(flowId, educationLevel, course, group, subgroup),
                lessonDate,
                lessonNum,
                willLessonBe,
                if (subjectId == 0L) null else Subject(subjectId, subject),
                if (teacherId == 0L) null else Teacher(teacherId, surname, name, patronymic),
                if (cabinetId == 0L) null else Cabinet(cabinetId, cabinet, building, address)
            )
        }
        cursor.close()
        db.close()
        return result
    }

    @SuppressLint("Range")
    fun findAllByFlow(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int
    ): List<TempSchedule> {
        var result = mutableListOf<TempSchedule>()

        val sql = """
            SELECT ts.id AS temp_id, ts.flow AS flow_id, ts.subject AS subject_id, ts.teacher AS teacher_id, ts.cabinet AS cabinet_id,
                ts.lesson_date, ts.lesson_num, ts.will_lesson_be,
                s.subject, t.surname, t.name, t.patronymic, c.cabinet, c.building, c.address
            FROM temp_schedule ts
            JOIN flow f ON ts.flow = f.id
            LEFT JOIN subject s ON ts.subject = s.id
            LEFT JOIN teacher t ON ts.teacher = t.id
            LEFT JOIN cabinet c ON ts.cabinet = c.id
            WHERE education_level=? AND course=? AND _group=? AND subgroup=? AND lesson_date=? AND lesson_num=?
        """.trimIndent()
        val selectionArgs = arrayOf(educationLevel.toString(), course.toString(), group.toString(), subgroup.toString())
        val db = dbHelper.readableDatabase
        val cursor: Cursor = db.rawQuery(sql, selectionArgs)
        if (cursor.moveToFirst()) {
            val flowId = cursor.getLong(cursor.getColumnIndex("flow_id"))

            val subjectId = cursor.getLong(cursor.getColumnIndex("subject_id"))
            val subject = cursor.getString(cursor.getColumnIndex("subject"))

            val teacherId = cursor.getLong(cursor.getColumnIndex("teacher_id"))
            val surname = cursor.getString(cursor.getColumnIndex("surname"))
            val name = cursor.getString(cursor.getColumnIndex("name"))
            val patronymic = cursor.getString(cursor.getColumnIndex("patronymic"))

            val cabinetId = cursor.getLong(cursor.getColumnIndex("cabinet_id"))
            val cabinet = cursor.getString(cursor.getColumnIndex("cabinet"))
            val building = cursor.getString(cursor.getColumnIndex("building"))
            val address = cursor.getString(cursor.getColumnIndex("address"))

            val id = cursor.getLong(cursor.getColumnIndex("temp_id"))
            val lessonDate = LocalDate.parse(cursor.getString(cursor.getColumnIndex("lesson_date")), DateTimeFormatter.ofPattern("yyyy-MM-dd"))
            val lessonNum = cursor.getInt(cursor.getColumnIndex("lesson_num"))
            val willLessonBe = cursor.getInt(cursor.getColumnIndex("will_lesson_be")) == 1

            result.add(TempSchedule(
                id,
                Flow(flowId, educationLevel, course, group, subgroup),
                lessonDate,
                lessonNum,
                willLessonBe,
                if (subjectId == 0L) null else Subject(subjectId, subject),
                if (teacherId == 0L) null else Teacher(teacherId, surname, name, patronymic),
                if (cabinetId == 0L) null else Cabinet(cabinetId, cabinet, building, address)
            ))
        }
        cursor.close()
        db.close()
        return result
    }

    fun deleteByFlowAndLessonDateAndLessonNum(
        educationLevel: Int,
        course: Int,
        group: Int,
        subgroup: Int,
        lessonDate: LocalDate,
        lessonNum: Int
    ): Int {
        val foundFlow: Flow = flowRepo.findByEducationLevelAndCourseAndGroupAndSubgroup(educationLevel, course, group, subgroup) ?: return 0

        val whereClause = "flow=? AND lesson_date=? AND lesson_num=?"
        val whereArgs = arrayOf(foundFlow.id.toString(), lessonDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")), lessonNum.toString())

        val db = dbHelper.writableDatabase
        val count = db.delete("temp_schedule", whereClause, whereArgs)
        db.close()
        return count
    }

    fun deleteAllBeforeDate(
        date: LocalDate
    ): Int {
        val whereClause = "lesson_date<?"
        val whereArgs = arrayOf(date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd")))

        val db = dbHelper.writableDatabase
        val count = db.delete("temp_schedule", whereClause, whereArgs)
        db.close()
        return count
    }
}
